// @nativescript/angular/esm2015/directives/list-view-comp.js
import { ChangeDetectionStrategy, Component, ElementRef, IterableDiffers, forwardRef, NgZone } from '@angular/core';
import { TEMPLATED_ITEMS_COMPONENT, TemplatedItemsComponent } from './templated-items-comp';
export class ListViewComponent extends TemplatedItemsComponent {
    constructor(_elementRef, _iterableDiffers, zone) {
        super(_elementRef, _iterableDiffers, zone);
    }
    get nativeElement() {
        return this.templatedItemsView;
    }
}
ListViewComponent.decorators = [
    { type: Component, args: [{
                selector: 'ListView',
                template: ` <DetachedContainer>
		<Placeholder #loader></Placeholder>
	</DetachedContainer>`,
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [{ provide: TEMPLATED_ITEMS_COMPONENT, useExisting: forwardRef(() => ListViewComponent) }]
            },] }
];
ListViewComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: IterableDiffers },
    { type: NgZone }
];
